### ISSUE TITLE
Publish to Appsource - Documentation

### ISSUE THREAD

**Issue by dsaveyn** [microsoft/AL-Go#979]
Is there any documentation available on releasing to Appsource? I found the relevant settings but it's a bit vague on how to configure it exactly. I'm especially struggling with the authentication. I already found this [blogpost](https://freddysblog.com/2022/09/22/automatic-appsource-submission-of-business-central-apps/) by googling, but a step-by-step guide would be useful.

How are library apps handled exactly? I have a main app that depends on a core app. I would expect the core app to be submitted as a library app for the main app. This is not the case however:

![image](https://github.com/microsoft/AL-Go/assets/92365904/2f350c46-a031-42ec-9433-97f2874bb5ad)

**Comment by dsaveyn** [comment]
@freddydk thank you this would help a lot. How is decided when a library app is supplied?

**Comment by dsaveyn** [comment]
Okay so I have been trying to figure out what's wrong with the AppSource submission  when it comes to library apps for the last couple of hours.

Our repo is structured as following (3 projects):

![image](https://github.com/microsoft/AL-Go/assets/92365904/d40e1fee-40e7-41d8-9913-24d8b324af99)

Core - app.json
```
{
  "id": "id-goes-here",         
  "name": "ESC Core",
  "publisher": "ESC bvba",
  "propagateDependencies": true,
  "dependencies": [
    {
      "id": "id-goes-here",
      "name": "Dynavision Core",
      "publisher": "Dynavision",
      "version": "23.0.4.0"
    }
  ]
}
```

Manufacturing Calculation - app.json

```
{
  "id": "id-goes-here",         
  "name": "Manufacturing Calculation",
  "publisher": "ESC bvba",
  "version": "23.0.0.0",
  "propagateDependencies": true,
  "dependencies": [
    {
      "id": "id-goes-here",
      "name": "ESC Core",
      "publisher": "ESC bvba",
      "version": "23.0.0.0"
    }    
  ]
}
```

A 23.0 release was created and it looks as following:

![image](https://github.com/microsoft/AL-Go/assets/92365904/3805ffa6-874e-4cc2-93d2-72259b4c5e2d)

When I try to create a release:

![image](https://github.com/microsoft/AL-Go/assets/92365904/f6d8cb24-5ba8-4048-8752-53ba30ec7c34)

The Core App is not included as a library app. 

I have been looking in the code and it appears that when the App Version To Deliver set to "current" and the Projects to Publish is set to "Manufacturing Calculation" , the latest release is downloaded. The projects are looped and the artifacts are filtered according to a mask "Apps". This means the Core App is never downloaded. When I set the Projects to Publish to "Core,Manufacturing Calculation" I get an error because the Core App is not an Appsource App.

So I'm a bit lost here. How do I include our Core App as a Library App in our submission?

I also noticed that in the artifacts the space is replaced by a dot for the project name:

```
Downloading release 23.0, projects Manufacturing Calculation, type Apps
project 'Manufacturing Calculation'
AssetPatterns: 'Manufacturing Calculation-*-Apps-*.zip' | 'Manufacturing Calculation-Apps-*.zip'
''
AL-Go action ran: Deliver Telemetry Correlation Id: 165ea514-35b6-4ce4-8954-d83bf5e97af7
Error: Unexpected error when running action. Error Message: Artifact current was not found on any release. Make sure that the artifact files exist and files are not corrupted., StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.1\Deliver\Deliver.ps1: line 113 <- at <ScriptBlock>, 
```

**Comment by dsaveyn** [comment]
Okay so I have been trying to figure out how the Publish to AppSource workflow works, but so far I has not been really succesful.

Our repo is structured as following (3 projects - 1 not relevant for this case):

![image](https://github.com/microsoft/AL-Go/assets/92365904/d40e1fee-40e7-41d8-9913-24d8b324af99)

Core - app.json
```
{
  "id": "id-goes-here",         
  "name": "ESC Core",
  "publisher": "ESC bvba",
  "propagateDependencies": true,
  "dependencies": [
    {
      "id": "id-goes-here",
      "name": "Dynavision Core",
      "publisher": "Dynavision",
      "version": "23.0.4.0"
    }
  ]
}
```

Manufacturing Calculation - app.json

```
{
  "id": "id-goes-here",         
  "name": "Manufacturing Calculation",
  "publisher": "ESC bvba",
  "version": "23.0.0.0",
  "propagateDependencies": true,
  "dependencies": [
    {
      "id": "id-goes-here",
      "name": "ESC Core",
      "publisher": "ESC bvba",
      "version": "23.0.0.0"
    }    
  ]
}
```

The Core App is a library for the Manufcaturing Calculation App

**A 23.0 release was created and it looks as following:**

![image](https://github.com/microsoft/AL-Go/assets/92365904/3805ffa6-874e-4cc2-93d2-72259b4c5e2d)

First **I tried to create a submission for the Manufacturing Calculation app with version "Latest"**:

![image](https://github.com/microsoft/AL-Go/assets/92365904/9efb5b02-a478-442a-8a4e-18607039bce6)

_A submission was created, but the Core was never included as a library app._

```
Artifacts latest
Projects:
Manufacturing Calculation
Project 'Manufacturing Calculation'
Finding latest successful CICD run for branch main in repository ESC-BV/bc-esc-appsource
- https://api.github.com/repos/ESC-BV/bc-esc-appsource/actions/runs?per_page=100&page=1&exclude_pull_requests=true&status=completed&branch=main
Last successful CICD run for branch main in repository ESC-BV/bc-esc-appsource is 82650[62](https://github.com/ESC-BV/bc-esc-appsource/actions/runs/8266253039/job/22613823414#step:5:64)496
Getting artifacts for workflow run 82[65](https://github.com/ESC-BV/bc-esc-appsource/actions/runs/8266253039/job/22613823414#step:5:67)062496, mask Apps, projects Manufacturing Calculation and version *
Found artifact Manufacturing Calculation-main-Apps-23.0.67.0 (ID: 1322283093) for mask Apps and project Manufacturing Calculation
No more artifacts found for workflow run 8265062496
Found 1 artifacts for mask Apps and projects Manufacturing Calculation in workflow run 8265062496
Downloading artifact Manufacturing Calculation-main-Apps-23.0.67.0
https://api.github.com/repos/ESC-BV/bc-esc-appsource/actions/artifacts/1322283093/zip
D:\a\bc-esc-appsource\bc-esc-appsource\.artifacts\Manufacturing Calculation-main-Apps-23.0.[67](https://github.com/ESC-BV/bc-esc-appsource/actions/runs/8266253039/job/22613823414#step:5:69).0.zip
Finding latest successful CICD run for branch main in repository ESC-BV/bc-esc-appsource
- https://api.github.com/repos/ESC-BV/bc-esc-appsource/actions/runs?per_page=100&page=1&exclude_pull_requests=true&status=completed&branch=main
Last successful CICD run for branch main in repository ESC-BV/bc-esc-appsource is 8265062496
Getting artifacts for workflow run 8265062496, mask Dependencies, projects Manufacturing Calculation and version *
No more artifacts found for workflow run 8265062496
Found 0 artifacts for mask Dependencies and projects Manufacturing Calculation in workflow run 8265062496
WARNING: Could not find any Dependencies artifacts for projects Manufacturing Calculation, version latest
Artifacts:
- Manufacturing Calculation-main-Apps-23.0.67.0
Applying settings from D:\a\bc-esc-appsource\bc-esc-appsource\.github\AL-Go-Settings.json
Applying settings from D:\a\bc-esc-appsource\bc-esc-appsource\Manufacturing Calculation\.AL-Go\settings.json
No settings found in D:\a\bc-esc-appsource\bc-esc-appsource\.github\Publish To AppSource.settings.json
No settings found in D:\a\bc-esc-appsource\bc-esc-appsource\Manufacturing Calculation\.AL-Go\Publish To AppSource.settings.json
No settings found in D:\a\bc-esc-appsource\bc-esc-appsource\Manufacturing Calculation\.AL-Go\dsaveyn.settings.json
Analyzing repository
Azure.Storage Module is available in version 2.1.0
Using Azure.Storage version 2.1.0
Attempting authentication to https://api.partner.microsoft.com/.default using clientCredentials...
Authenticated as app ***
AppSource MainAppFolder App
Main App File:
- ESC bvba_Manufacturing Calculation_23.0.67.0.app
Library App Files:
- None
Submitting to AppSource
```

To figure out what was wrong I opened the AL-Go source and noticed that current uses the latest build and not the latest release. 

**Therefore I tried to create a new submission for the Manufacturing Calculation app with version "Current":**

![image](https://github.com/microsoft/AL-Go/assets/92365904/dc4c9eb5-6dc9-406f-8ab2-7db4e590defa)

_Now the submission failed:_

```
Artifacts current
Projects:
Manufacturing Calculation
Project 'Manufacturing Calculation'
Analyzing releases https://api.github.com/repos/ESC-BV/bc-esc-appsource/releases
Downloading release 23.0, projects Manufacturing Calculation, type Apps
project 'Manufacturing Calculation'
AssetPatterns: 'Manufacturing Calculation-*-Apps-*.zip' | 'Manufacturing Calculation-Apps-*.zip'
''
AL-Go action ran: Deliver Telemetry Correlation Id: 165ea514-35b6-4ce4-89[54](https://github.com/ESC-BV/bc-esc-appsource/actions/runs/8267190719/job/22617020948#step:5:56)-d83bf5e97af7
Error: Unexpected error when running action. Error Message: Artifact current was not found on any release. Make sure that the artifact files exist and files are not corrupted., StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.1\Deliver\Deliver.ps1: line 113 <- at <ScriptBlock>, D:\a\_temp\00bd3298-0ad6-497a-8180-1fd22ce91cf9.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1
Error: Process completed with exit code 1.
```

So it appears the code is looking for Manufacturing Calculation while the artifacts are named as Manufacturing.Calculation. So it appears the space is replaced by a dot in the project name. **==> Bug?**

I did some further digging but it's unclear to me how the Core App should be included as a library app when the version is set to "Current": the latest release is downloaded, the projects are looped and the artifacts are filtered according to a mask "Apps". 
This means the Core App will never be downloaded. I tried including the Core App in the apps to publish "Core,Manufacturing Calculation" but then I get an error because the Core App is not an AppSource App.

So I'm a bit lost here. How do I include our Core App as a Library App in our submission?

**Comment by freddydk** [comment]
Thanks for linking this to 1009 - I will add the mentioned guide to the fix for this.
